### ISSUE TITLE
Debugger crashes when evaluating watch variables on Elixir 1.10

### ISSUE THREAD

**Issue by cec** [KronicDeth/intellij-elixir#1659]
**Describe the bug**
Evaluating new watch expression in debugger crashes the debugger.

**To Reproduce**
1. Start a Debug configuration 
2. Place a breakoint
3. Have the debugger hit the breakpoint
4. Click the + 'Add watch' button
5. Type a valid elixir expression to watch
6. See the debugger crash

**Expected behavior**
Watch expressions should be evaluates successfully and not crash the debugger.

**Logs**
````
[error] GenServer IntelliJElixir.Debugger.Server terminating
** (FunctionClauseError) no function clause matching in :elixir_env.prepare_write/1
    (elixir 1.10.0) src/elixir_env.erl:71: :elixir_env.prepare_write(#Macro.Env<aliases: [], context: nil, context_modules: [], file: "/home/cec/core/learn/elixir/phoenix/edu_bookings/lib/edu_bookings_web/controllers/venue_controller.ex", function: {:index, 2}, functions: [{Kernel, [!=: 2, !==: 2, *: 2, ...]}], lexical_tracker: nil, line: 11, macro_aliases: [], macros: [{Kernel, ...}], module: EduBookingsWeb.VenueController, requires: [...], ...>)
    (elixir 1.10.0) src/elixir_expand.erl:402: :elixir_expand.expand/2
    /tmp/intellij_elixir7/debugger/lib/intellij_elixir/debugger/server.ex:398: IntelliJElixir.Debugger.Server.quoted_to_erl/3
    /tmp/intellij_elixir7/debugger/lib/intellij_elixir/debugger/server.ex:277: IntelliJElixir.Debugger.Server.handle_call/3
    (stdlib 3.11.2) gen_server.erl:661: :gen_server.try_handle_call/4
    (stdlib 3.11.2) gen_server.erl:690: :gen_server.handle_msg/6
    (stdlib 3.11.2) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Last message (from #PID<7476.3.0>): {:evaluate, %{env: %{file: "/home/cec/core/learn/elixir/phoenix/edu_bookings/lib/edu_bookings_web/controllers/venue_controller.ex", function: {:index, 2}, line: 11, module: EduBookingsWeb.VenueController}, expression: "length(venues)", pid: #PID<0.6733.0>, stack_pointer: 8}}
State: %IntelliJElixir.Debugger.Server{attached: #PID<7476.3.0>, evaluate_meta_pid_to_froms: %{}}
Client #PID<7476.3.0> is remote on node :"debugger62a32658-0206-4ce6-82f4-523ff9cb93c6@127.0.0.1"
````

**Desktop:**
 - OS: [Linux Ubuntu]
 - Version [18.04 LTS]

**Erlang:**
- Installer [apt repository (deb https://packages.erlang-solutions.com/ubuntu bionic contrib) ]
- Version [22]

**Elixir:**
- Installer [apt repository (deb https://packages.erlang-solutions.com/ubuntu bionic contrib) ]
- Version [1.10.0]

**Plugin:**
- Version [11.5.0]

**Additional context**

**Comment by cec** [comment]
I guess a change in 1.10 on the structure of the params expected by `:elixir_env.prepare_write/1` is to blame?

**Comment by KronicDeth** [comment]
Yes, the debugger API is unstable.  I have to adapt to any changes the core team made to it for IEx. Pry to keep the evaluator working.  If you can find the commit/PR that changed Elixir I can figure out how to port it to the debugger in the plugin.  Otherwise, I’ll search for it myself later.

**Comment by cec** [comment]
Hi :)

https://github.com/elixir-lang/elixir/commit/60a472e1ae288c867ee93d3e834ed8d612a4d3bc seems to be the "culprit"

**Comment by KronicDeth** [comment]
@cec the debugging API is non-public.  As José has put it, I can use it, but not one else can.  I think in the sense that I'm the only one nuts enough to keep up with the changes.

**Comment by sh41** [comment]
Your guess was right, and 60a472e is the right commit — though not quite because 1.10 changed what
`:elixir_env.prepare_write/1` expects. That commit is where `prepare_write/1` was *added*, and it first
shipped in 1.10.0, which is why your trace names it. The shape it wants — `current_vars` holding a
two-element read/write tuple — is what Elixir used on 1.7 through 1.9 as well.

The badly shaped value is ours. To evaluate an expression inside an interpreted process the debugger
can't use `Code.eval_quoted`, so it rebuilds what IEx does out of Elixir's private compiler internals.
It has `:elixir_env.with_vars/2` build the environment, which shapes `current_vars` correctly, and then
immediately overwrites that field with a plain map of the watch variables. `prepare_write/1`
pattern-matches the tuple, a plain map doesn't match, and you get the FunctionClauseError.

Elixir 1.13 renamed that field to `versioned_vars`. The overwrite then matched nothing and quietly
stopped happening, which is what made the crash go away — the code is unchanged, the structure it was
corrupting moved out from under it. The oldest Elixir this project's CI tests against is now 1.13.4, so
no supported configuration can still reach it.

Closing on that basis. #1858 is the same crash reported against 1.11.2 and was closed the same way. The
wider problem — the evaluate path reconstructing private compiler internals behind version checks that
stop at Elixir 1.13, while current Elixir is 1.20 — is real, unresolved, and tracked in #3944. If you
see an evaluate crash on a recent Elixir, that is worth a new report rather than a reopen here, since it
will be a different mechanism.

Closed as part of an AI-assisted review of the backlog. A human checked this one before closing it, but
it might still be a mistake. If you think the reason above is wrong, please reopen the issue and tell us
why.
